### ISSUE TITLE
Neglected machine learning ideas | Locklin on science

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4973]
Neglected machine learning ideas | Locklin on science<br>
<br>
https://ift.tt/1AH7qKO<br>
<br>
<br>
<br>
<div>
<div>
<p>This post is inspired by the &ldquo;metacademy&rdquo; suggestions for <a href="http://metacademy.org/roadmaps/cjrd/level-up-your-ml">&ldquo;leveling up your machine learning</a>.&rdquo; They make some halfway decent suggestions for beginners.&nbsp; The problem is, these suggestions won&rsquo;t give you a view of machine learning as a field; they&rsquo;ll only teach you about the subjects of interest to authors of machine learning books, which is different. The level-3 and level-4 suggestions they make are not super useful either: they just reflect the tastes of the author.</p>
<p>The machine learning literature is vast, techniques are bewilderingly diverse, multidisciplinary and seemingly unrelated. It is extremely difficult to know what is important and useful. While &ldquo;metacademy&rdquo; has the horse sense to suggest reading some books, the problem is, there is no book which can even give you a survey of what is available, or make you aware of things which might be helpful. The best guide for the perplexed, in my not at all humble opinion, is Peter Flach&rsquo;s introductory text, &ldquo;<a href="http://www.amazon.com/Machine-Learning-Science-Algorithms-Sense/dp/1107422221">Machine Learning: the Art and Science of Algorithms that Make Sense of Data</a>&rdquo; which at least mentions some of the more obscure techniques, and makes pointers to other resources. Most books are just a collection of the popular techniques. They all mention regression models, logistic regression, neural nets, trees, ensemble methods, graphical models and SVM type things. Most of the time, they don&rsquo;t even bother telling you what each technique is actually good for, and when you should choose one over the other for an approach (Flach does; that&rsquo;s one of many reasons you should read his book). Sometimes I am definitely just whining that people don&rsquo;t pay enough attention to the things I find interesting, or that I don&rsquo;t have a good book or review article on the topic. Sleep deprivation will do that to a man. Sometimes I am probably putting together things that have no clearly unifying feature, perhaps because they&rsquo;re &ldquo;not done yet.&rdquo; I figure that&rsquo;s OK, subjects such as &ldquo;deep learning&rdquo; are also a bunch of ideas that have no real unifying theme and aren&rsquo;t done yet; this doesn&rsquo;t stop people from writing <a href="http://deeplearning.net/">good treatments</a> of the subject. Perhaps my list is a &ldquo;send me review articles and book suggestions&rdquo; cry for help, but perhaps it is useful to others as an overview of neat things.</p>
<p><img alt="1510" src="https://scottlocklin.files.wordpress.com/2014/07/1510.jpg?w=300&amp;h=187"></p>
<p>&nbsp;</p>
<p><strong><em>Stuff I think is egregiously neglected in books, and in academia in unranked semi-clustered listing below:</em></strong></p>
<p>&nbsp;</p>
<p><strong>Online learning:</strong> not the &ldquo;Khan academy&rdquo; kind, the &ldquo;<a href="https://en.wikipedia.org/wiki/Online_machine_learning">exposing your learners to data, one piece at a time</a>, the way the human&nbsp; brain works&rdquo; kind. This is hugely important for &ldquo;big data&rdquo; and timeseries, but there are precious few ML texts which go beyond mentioning the existence of online learning in passing. Almost all textbooks concentrate on batch learning. Realistically, when you&rsquo;re dealing with timeseries or very large data sets, you&rsquo;re probably doing things online in some sense. If you&rsquo;re not thinking about how you&rsquo;re exposing your learners to sequentially generated data, you&rsquo;re probably leaving information on the table, or overfitting to irrelevant data. I can think of zero books which are actually helpful here. Cesa-Bianchi and Lugosi wrote a <a href="http://www.amazon.com/Prediction-Learning-Games-Nicolo-Cesa-Bianchi/dp/0521841089">very interesting book</a> on some recent proofs for online learners and &ldquo;<a href="http://vserver1.cscs.lsa.umich.edu/~crshalizi/notabene/universal-prediction.html">universal prediction</a>&rdquo; which strike me as being of extreme importance, though this is a presentation of new ideas, rather than an exposition of established ones.&nbsp; <a href="http://hunch.net/~vw/">Vowpal Wabbi</a>t is a useful and interesting piece of software with OK documentation, but there should be a book which takes you from online versions of linear regression (they exist! I can show you <a href="http://queue.acm.org/detail.cfm?id=2507780">one</a>!)&nbsp; to something like Vowpal Wabbit. Such a book does not exist. Hell, I am at a loss to think of a decent review article, and the subject is unfortunately un-googleable, thanks to the hype over the BFD of &ldquo;watching lectures and taking tests over the freaking internets.&rdquo; Please correct me if I am wrong: I&rsquo;d love to have a good review article on the subject for my own purposes.</p>
<p><img alt="boris-artzybasheff-machinalia1" src="https://scottlocklin.files.wordpress.com/2014/07/boris-artzybasheff-machinalia1.jpg?w=300&amp;h=202"></p>
<p><strong>Reinforcement learning</strong>: a <a href="https://en.wikipedia.org/wiki/Reinforcement_learning">form</a> of online learning which has become a field unto its own. One of the great triumphs of machine learning is teaching computers to win at <a href="https://en.wikipedia.org/wiki/TD-Gammon">Backgammon</a>. This was done via a form of reinforcement learning known as <a href="https://en.wikipedia.org/wiki/Temporal_difference_learning">TD-learning</a>. Reinforcement learning is a large field, as it has been used with great success in control systems theory and robotics. The problem is, the guys who do reinforcement learning are generally in control systems theory and robotics, making the literature impenetrable to machine learning researchers and engineers. Something oriented towards non robotics problems would be nice (Sutton and Barto doesn&rsquo;t suffice here; Norvig&rsquo;s chapter is the best general treatment I have thus far seen). There are papers on applications of the idea to ideas which do not involve robots, but none which unify the ideas into something comprehensible and utile to a ML engineer.</p>
<p><img alt="4140093417_c8b498a205_b" src="https://scottlocklin.files.wordpress.com/2014/07/4140093417_c8b498a205_b.jpg?w=300&amp;h=203"></p>
<p><strong>&ldquo;Compression&rdquo; sequence prediction techniques</strong>: this is another form of online learning, though it can also be done in batch mode. We&rsquo;re all familiar with this; when google tries to guess what you&rsquo;re going to search for, it is using a primitive form of this called the <a href="https://en.wikipedia.org/wiki/Trie">Trie</a>. Such ideas are related to standard compression techniques like LZW, and have deep roots in information theory and signal processing. Really, <a href="https://scottlocklin.wordpress.com/2013/10/31/shannon-information-compression-psychic/">Claude Shannon</a> wrote the first iterations of this idea. I can&rsquo;t give you a good reference for this subject in general, though <a href="http://www.cs.technion.ac.il/~ronbeg/vmm/index.html">Ron Begleiter and friends</a> wrote a very good paper on some classical compression learning implementations and their uses. I wrote an <a href="https://github.com/locklin/VMM">R wrapper</a> for their Java lib if you want to fool around with their tool. Boris Ryabko and son have also written <a href="http://boris.ryabko.net/papers.html">numerous interesting papers</a> on the subject. <a href="http://www.complearn.org/">Complearn</a> is a presumably useful library which encapsulates some of these ideas, and is available everywhere Linux is sold. Some day I&rsquo;ll expound on these ideas in more detail.</p>
<p><img alt="4140853224_78a8faed00_b" src="https://scottlocklin.files.wordpress.com/2014/07/4140853224_78a8faed00_b.jpg?w=300&amp;h=203"></p>
<p><strong>Time series oriented techniques in general</strong>: a large fraction of&nbsp; industry&nbsp;applications have a time component. Even in marketing problems dealing with survival techniques, there is a time component, and you should know about it.In situations where there are non-linear relationships in the time series, <a href="https://en.wikipedia.org/wiki/Autoregressive_integrated_moving_average">classical</a> regression and time-series techniques will fail. In situations where you must discover the underlying <a href="https://en.wikipedia.org/wiki/Extended_Kalman_filter">non-linear model yourself</a>, well, you&rsquo;re in deep shit if you don&rsquo;t know some time-series oriented machine learning techniques.&nbsp; There was much work done in the 80s and 90s on tools like <a href="https://en.wikipedia.org/wiki/Recurrent_neural_network">recurrent ANNs</a> and <a href="https://en.wikipedia.org/wiki/Feedforward_neural_network">feedforward ANNs</a> for starters, and there has been much work in this line since then. There are plenty of other useful tools and techniques.&nbsp; Once in a while someone will mention <a href="http://cran.r-project.org/web/packages/dtw/index.html">dynamic time warping</a> in a book, but nobody seems real happy about this technique.&nbsp; Many books mention Hidden Markov Models, which are important, but they&rsquo;re only useful when the data is at least semi-Markov, and you have some idea of how to characterize it as a sequence of well defined states. Even in this case, I daresay not even the natural language recognition textbooks are real helpful (though <a href="http://www.amazon.com/Fundamentals-Speech-Recognition-Lawrence-Rabiner/dp/0130151572">Rabiner and Juang</a> is OK, it&rsquo;s also over 20 years old). Similarly, there are no review papers&nbsp; treating this as a general problem. I guess we TS guys are too busy racking in the lindens to write one.</p>
<p><strong>Conformal prediction</strong>: I will be surprised if anyone reading this has even heard of <a href="http://www.vovk.net/cp/">conformal prediction</a>. There are no wikipedia entries. There is a <a href="http://onlineprediction.net/?n=Main.ConformalPrediction">website</a> and a <a href="http://www.alrw.net/">book</a>. The concept is simple: it would be nice to well motivated put error bars on a machine learning prediction. If you read the basic books, stuff like k-fold cross validation and the <a href="https://en.wikipedia.org/wiki/Jackknife_resampling">jackknife</a>&nbsp; trick are the entire story. OK, WTF do I do when my training is online? What do I do in the presence of different kinds of noise? Conformal prediction is a step towards this, and hopefully a theory of machine learning confidence intervals in general. It seems to mostly be the work of a <a href="http://www.glennshafer.com/">small</a> group of <a href="http://www.vovk.net/">researchers</a> who were influenced by Kolomogorov, but <a href="http://arxiv.org/find/all/1/all:+EXACT+conformal_prediction/0/1/0/all/0/1">others</a> are catching on. I&rsquo;m interested. Not interested enough to write one, as of yet, but I&rsquo;d sure like to play with one.</p>
<p><img alt="98" src="https://scottlocklin.files.wordpress.com/2014/07/98.jpg?w=300&amp;h=205"></p>
<p><strong>ML in the presence of lots of noise:</strong> The closest thing to a book on it is the bizarro (and awesomely cool) &ldquo;<a href="http://www.amazon.com/Pattern-Theory-Stochastic-Real-World-Mathematics/dp/1568815794">Pattern Theory: The Stochastic Analysis of Real World Signals</a>&rdquo; by Mumford and Desolneux, or perhaps something in the corpus of speech recognition and image processing books. This isn&rsquo;t exactly a cookbook or exposition, mind you: more of a thematic manifesto with a few applications.&nbsp; Obviously, signal processing has something to say about the subject, but what about <em>learners</em> which are <em>designed</em> to function usefully when we know that most of the data is noise?&nbsp; Fields such as natural language processing and image processing are effectively ML in the presence of lots of noise and confounding signal, but the solutions you will find in their textbooks are specifically oriented to the problems at hand.&nbsp; Once in a while something like vector quantization will be reused across fields, but it would be nice if we had an &ldquo;elements of statistical learning in the presence of lots of noise&rdquo; type book or review paper. Missing in action, and other than the specific subfields mentioned above, there are no research groups which study the problem as an engineering subject. New stuff is happening all the time; part of the success of &ldquo;Deep Learning&rdquo; is attributable to the <a href="http://nips.cc/Conferences/2012/Program/event.php?ID=3685">Drop Out technique</a> to prevent overfitting. Random forests could be seen as a technique which at genuflects at&nbsp; &ldquo;ML in the presence of noise&rdquo; without worrying about it too much. Marketing guys are <a href="https://scottlocklin.wordpress.com/2014/07/22/neglected-machine-learning-ideas/scientificmarketer.com/2007/02/modelling-that-which-is-important.html">definitely</a> thinking about this. I know for a fact that there are very powerful learners for picking signal out of shitloads of noise: I&rsquo;ve written some. It would have been a lot easier if somebody wrote a&nbsp; review paper on the topic. The available knowledge can certainly be systematized and popularized better than it has been.</p>
<p><img alt="4140852076_c0b2bc6ee6_b" src="https://scottlocklin.files.wordpress.com/2014/07/4140852076_c0b2bc6ee6_b.jpg?w=300&amp;h=198"></p>
<p><strong>Feature engineering</strong>: feature engineering is another topic which doesn&rsquo;t seem to merit any review papers or books, or even chapters in books, but it is absolutely vital to ML success. Sometimes the features are obvious; sometimes not. Much of the success of machine learning is actually success in engineering features that a learner can understand. I daresay document classification would be awfully difficult without <a href="https://en.wikipedia.org/wiki/Tf%E2%80%93idf">td-idf representation</a> of document features. <del><a href="https://en.wikipedia.org/wiki/Latent_Dirichlet_allocation">Latent Dirichlet allocation</a> is a form of &ldquo;graphical model&rdquo; which works wonders on such data, but it wouldn&rsquo;t do a thing without td-idf.</del> [correction to this statement from <a href="https://scottlocklin.wordpress.com/2014/07/22/neglected-machine-learning-ideas/#comment-8805">Brendan below</a>] Similarly, image processing has a bewildering variety of feature extraction algorithms which are of towering importance for that field; the <a href="https://en.wikipedia.org/wiki/Scale-invariant_feature_transform">SIFT</a> descriptor, the <a href="https://www.quora.com/Computer-Vision/What-is-a-GIST-descriptor">GIST</a> and <a href="http://www.vlfeat.org/overview/hog.html">HOG</a> descriptors, the <a href="https://en.wikipedia.org/wiki/Hough_transform">Hough</a> transform, <a href="http://www.data-compression.com/vq.shtml">vector quantization</a>, tangent distance [<a href="https://scottlocklin.wordpress.com/2014/07/22/neglected-machine-learning-ideas/yann.lecun.com/exdb/publis/pdf/simard-00.pdf">pdf link</a>]. The Winner Take All hash [<a href="http://www.robots.ox.ac.uk/~vgg/rg/papers/37298.pdf">pdf link</a>] is an extremely <a href="https://gist.github.com/erogol/6391423">simple</a> and related idea&hellip; it makes a man wonder if such ideas could be used in higher (or lower) dimensions. Most of these engineered features are histograms in some sense, but just saying &ldquo;use a histogram&rdquo; isn&rsquo;t helpful. A review article or a book chapter on this sort of thing, thinking through the relationships of these ideas, and helping the practitioner to engineer new kinds of feature for broad problems would be great. Until then, it falls to the practitioner to figure all this crap out all by their lonesome.</p>
<p><img alt="4140092567_80f07dbc23_b" src="https://scottlocklin.files.wordpress.com/2014/07/4140092567_80f07dbc23_b.jpg?w=300&amp;h=197"></p>
<p><strong>Unsupervised and semi-supervised learning in general</strong>: almost all books, and even tools like R inherently assume that you are doing supervised learning, or else you&rsquo;re doing something real simple, like hierarchical clustering, kmeans or PCA.&nbsp; In the presence of a good set of features, or an interesting set of data, <a href="https://en.wikipedia.org/wiki/Unsupervised_learning">unsupervised techniques</a> can be very helpful. Such techniques may be crucial. They may even help you to engineer new features, or at least reduce the dimensionality of your data. Many interesting data sets are only possible to analyze using <a href="https://en.wikipedia.org/wiki/Semi-supervised_learning">semi-supervised techniques</a>; recommendation engines being an obvious beneficiary of such tricks. &ldquo;Deep learning&rdquo; is also connected with unsupervised and semi-supervised approaches. I am pretty sure the genomics community does a lot of work with this sort of thing for dimensionality reduction. Supposedly <a href="http://cran.r-project.org/web/packages/rgp/index.html">Symbolic Regression</a> (generalized additive models picked using genetic algorithms) is pretty cool too, and it&rsquo;s in my org-emacs TODO lists to look at this more. Lots of good unsupervised techniques such as Kohonen <a href="https://en.wikipedia.org/wiki/Self-organizing_map">Self Organizing Maps</a> have fallen by the wayside. They&rsquo;re still useful: I use them. I&rsquo;d love a book or review article which concentrates on the topic, or just provides a bestiary of things which are broadly unsupervised. I suppose <a href="http://olivier.chapelle.cc/ssl-book/">Oliver Chapelle&rsquo;s book</a> is an OK start for semi-supervised ideas, but again, not real unified or complete.</p>
<p><em>&nbsp;Images by one of my heroes, the Ukrainian-American artist <a href="https://en.wikipedia.org/wiki/Boris_Artzybasheff">Boris Artzybasheff</a>. You can find more of it <a href="http://www.darkroastedblend.com/2014/01/machines-alive-whimsical-art-of-boris.html">here</a>.</em></p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Locklin on science https://ift.tt/1fNQBnY<br>
<br>
August 1, 2019 at 04:45PM
